Maze.cpp

```diff
@@ -34,11 +38,194 @@ void Maze::setWall(CellCoordinate pos, Direction dir, bool wall) {
+	
+	// The start node is 0 distance away.
+	startNode->setScores(0, heuristic(start, end));
+	startNode->direction = facing;
+
+	// While there are node remaining to be evaluated.
+	while (!openNodes.empty())
+	{
+		// Get the open node with lowest score.
+		auto currentNodeItr = openNodes.begin();
+
+		// If the current node is the finish node.
+		if ( *currentNodeItr == getNode( end ) )
+		{
+			// We are done. Construct the path.
+			return constructPath(*currentNodeItr);
+		}
+		
+		// Set the current node as evaluated.
+		openNodes.erase(currentNodeItr);
+		(*currentNodeItr)->evaluated = true;
+
+		// For each node adjacent to the current node.
+		for (
+			Direction direction = Direction::N;
+			direction != Direction::NONE;
+#ifdef MAZE_DIAGONALS
+			direction = static_cast<Direction>(direction + 1)
+#else
+			direction = static_cast<Direction>(direction + 2)
+#endif // MAZE_DIAGONALS
+			)
+		{
+			Node * adjacentNode = getAdjacentNode(*currentNodeItr, direction);
+
+			// If the adjacent node has already been evaluated or does not exist.
+			if (adjacentNode->evaluated || !adjacentNode->exists)
+			{
+				// Ignore the adjacent node.
+				continue;
+			}
+
+			auto tentativeScore = (*currentNodeItr)->gScore + calculateMovementCost( (*currentNodeItr)->direction, direction );
+
+			// If adjacent node not in openNodes.
+			if (openNodes.find(adjacentNode) == openNodes.end() )
+			{
+				// Discover a new node.
+				openNodes.insert(adjacentNode);
+			}
+			// If the previous score given to the adjacent node is better.
+			else if (tentativeScore >= adjacentNode->gScore)
+			{
+				// This is not a better path.
+				continue;
+			}
+
+			// This path is the best so far.
+			adjacentNode->previous = *currentNodeItr;
+			auto heuristicScore = heuristic((*currentNodeItr)->pos, end);
+			(*currentNodeItr)->setScores(tentativeScore, adjacentNode->gScore + heuristicScore);
+		}
+	}
+
+	throw runtime_error("No path found");
+}
+
+Node * Maze::getNode(NodeCoordinate pos)
 {
+	return &maze[pos.y][pos.x];
+}
 
+Node * Maze::getAdjacentNode(Node * node, Direction direction)
+{
+	return getNode(node->pos + direction);
+}
+
+void Maze::resetNodePathData()
+{
+	for (size_t y = 0; y < CELL_ROWS; y++)
+	{
+		for (size_t x = 0; x < CELL_COLS; x++)
+		{
+			Node * node = getNode(NodeCoordinate(x, y));
+
+			if (*node)
+			{
+				node->resetPathData();
+			}
+		}
+	}
+}
+
+unsigned Maze::calculateMovementCost(Direction currentDirection, Direction nextDirection)
+{
+	unsigned cost;
+	switch (nextDirection)
+	{
+	case N:
+	case E:
+	case S:
+	case W:
+		cost = MOVEMENT_COST;
+	case NE:
+	case SE:
+	case SW:
+	case NW:
+		cost = MOVEMENT_COST_DIAGONAL;
+	default:
+		throw std::invalid_argument("Direction cannot be: NONE");
+	}
+
+	// We are not given what direction we are currently facing so no penalty can be given.
+	if (currentDirection == NONE)
+	{
+		return cost;
+	}
+
+	// Penalize turns.
+	switch ( abs(nextDirection - currentDirection) )
```

Comment:
I don't know. It might be better to just use an integer, although it would only take a discreet number of values.